plugin-tooling.mk

```diff
@@ -0,0 +1,97 @@
+PLUGIN_BUILD_DATE ?= $(shell date -u +"%Y-%m-%d")
+PLUGIN_BUILD_VERSION ?= $(shell git describe --tags 2>$(NUL))
+
+ifeq ($(strip $(PLUGIN_BUILD_VERSION)),)
+PLUGIN_BUILD_VERSION = dev
+endif
+PLUGIN_LD_FLAGS += -X 'github.com/vmware-tanzu/tanzu-plugin-runtime/plugin/buildinfo.Date=$(PLUGIN_BUILD_DATE)'
+PLUGIN_LD_FLAGS += -X 'github.com/vmware-tanzu/tanzu-plugin-runtime/plugin/buildinfo.SHA=$(PLUGIN_BUILD_SHA)'
+PLUGIN_LD_FLAGS += -X 'github.com/vmware-tanzu/tanzu-plugin-runtime/plugin/buildinfo.Version=$(PLUGIN_BUILD_VERSION)'
+
+# Add supported OS-ARCHITECTURE combinations here
+PLUGIN_BUILD_OS_ARCH ?= linux-amd64 windows-amd64 darwin-amd64
+
+# Paths and Directory information
+ROOT_DIR := $(shell git rev-parse --show-toplevel)
+
+PLUGIN_DIR := ./cmd/plugin
+PLUGIN_BINARY_ARTIFACTS_DIR := $(ROOT_DIR)/artifacts/plugins
+PLUGIN_PACKAGE_ARTIFACTS_DIR := $(ROOT_DIR)/artifacts/packages
+
+REGISTRY_PORT := 5001
+REGISTRY_ENDPOINT := localhost:$(REGISTRY_PORT)
+PLUGIN_NAME := "*"
+
+# Repository specific configuration
+BUILDER := $(ROOT_DIR)/bin/builder
+PUBLISHER := tkg
+VENDOR := vmware
+PLUGIN_PUBLISH_REPOSITORY := localhost:$(REGISTRY_PORT)/test/v1/tanzu-cli/plugins
+
+## --------------------------------------
+## Plugin Build and Publish Tooling
+## --------------------------------------
+
+PLUGIN_BUILD_TARGETS := $(addprefix plugin-build-,${PLUGIN_BUILD_OS_ARCH})
+
+.PHONY: plugin-build
+plugin-build: $(PLUGIN_BUILD_TARGETS) ## Build all plugin binaries for all supported os-arch
+
+plugin-build-local: plugin-build-$(GOHOSTOS)-$(GOHOSTARCH) ## Build all plugin binaries for local platform
+	
+plugin-build-%:
+	$(eval ARCH = $(word 2,$(subst -, ,$*)))
+	$(eval OS = $(word 1,$(subst -, ,$*)))
+	$(BUILDER) plugin build \
+		--path $(PLUGIN_DIR) \
+		--binary-artifacts $(PLUGIN_BINARY_ARTIFACTS_DIR) \
+		--version $(PLUGIN_BUILD_VERSION) \
+		--ldflags "$(PLUGIN_LD_FLAGS)" \
+		--os-arch $(OS)_$(ARCH) \
+		--match $(PLUGIN_NAME)
+
+.PHONY: plugin-build-packages
+plugin-build-packages: local-registry ## Build plugin packages
+	$(BUILDER) plugin build-package \
+		--binary-artifacts $(PLUGIN_BINARY_ARTIFACTS_DIR) \
+		--package-artifacts $(PLUGIN_PACKAGE_ARTIFACTS_DIR) \
+		--oci-registry $(REGISTRY_ENDPOINT)
+
+.PHONY: plugin-publish-packages
+plugin-publish-packages: ## Publish plugin packages
+	$(BUILDER) plugin publish-package \
+		--package-artifacts $(PLUGIN_PACKAGE_ARTIFACTS_DIR) \
+		--publisher $(PUBLISHER) \
+		--vendor $(VENDOR) \
+		--repository $(PLUGIN_PUBLISH_REPOSITORY)
+
+.PHONY: plugin-combined-build-publish-packages
+plugin-combined-build-publish-packages: plugin-build plugin-build-packages plugin-publish-packages ## Build and Publish plugin packages
+
+## --------------------------------------
+## docker
+## --------------------------------------
+
+.PHONY: local-registry
+local-registry: clean-registry ## Starts up a local docker registry for the e2e tests
```

Comment:
nit: 

- The comment should be updated since this target is no longer just for e2e but is also needed to bring up the registry for generating 
- Any container with same name is going to be killed before a new one is started. Consider adding a comment header at least to warn about this behavior.
- To minimize surprises, maybe we can avoid using generic container name like "registry", and instead using something that makes it obvious it is a temp registry started solely to push and save packages during build.